### PR TITLE
tooling: generate JSON schemas from protobuf definitions

### PR DESCRIPTION
**Pull request by norbjd**
**Commit Message**

Generate JSON schemas from protobuf definitions. This is mostly useful to validate syntax when writing the config, using for example IDE extensions (the most well known is probably [redhat's vscode-yaml one](https://github.com/redhat-developer/vscode-yaml)).

**Additional Description**

The process to write correct envoy YAML configuration files feels inefficient. As of now, there is no way to validate the syntax while editing the config - except raw YAML validation - so the process (at least for me) is basically always: 1) write config file and 2) run `validate` command using an Envoy Docker image (with the version I want). As a result, the developer  feedback loop is slow.

This PR is an attempt to generate JSON schemas from protobuf definitions, so that they could be included in our favorite IDEs for validation. There might be other use cases I have not in mind right now that could now be possible with these.

Some examples of what it looks like after generating and using the JSON schemas (in [vscode](https://github.com/microsoft/vscode)).

*Autocompletion*

Fields:

image

Values (interestingly, autocompletion for enums like the following also gives numerical values, but it works!):

image

*Syntax validation (here, typing)*

image

----

Note that there are already a few issues mentioning this:

- https://github.com/envoyproxy/envoy/issues/13078
- https://github.com/envoyproxy/envoy/issues/13233
- https://github.com/envoyproxy/envoy/issues/13254

And some attempts to solve them:

- https://github.com/jcchavezs/envoy-config-schema

There is even one draft PR on Envoy:

- https://github.com/envoyproxy/envoy/pull/15100

However, there is no activity on these ones for at least one year (and often more) as of today.

The main reason and the main blocker for this was, I think, the fact that it was not really easy to integrate the actual plugin to generate JSON schemas from protobuf files (https://github.com/chrusty/protoc-gen-jsonschema). This is mostly because this external repository was not "bazelified". So, in order to reduce the amount of code in Envoy codebase and to simplify, I've decided to first add Bazel support to that plugin: https://github.com/chrusty/protoc-gen-jsonschema/compare/master...norbjd:add-bazel, just for Envoy. If everything works here with my PR on Envoy, then I'll create the PR to merge upstream and contribute to chrusty's repository.

Once done, it's now "pretty" easy to generate JSON schemas from Envoy protobuf files. I'm putting "pretty" between quotes because I'm not a Bazel expert, and it looks like there are many conventions for Bazel in Envoy that I don't fully master. But since I had some bandwidth to work on that, I'm happy to be able to contribute! I still have a few questions, and will open a comment in this PR for everyone of them, so that Bazel masters could give me some advice. I'm also available on [Slack](https://envoyproxy.slack.com) (my username is `norbjd`).

Finally, there is still one remaining question about the expected output. Ideally, we'd like to have a single JSON schema file (or two maybe, one for v2 and one for v3), but the protoc plugin does not handle this, so some work may need to be done in Envoy repository to merge all generated JSON schemas together. This/these final JSON files could be exposed on Envoy website, or in a first step, just downloadable from Github releases artifacts. I guess it's something that could be done in a second step to not overload this PR and avoid scope creep.

Below you can find commands to test the feature.

*1. Setup a clean environment with docker*

```shell
git clone -b generate-json-schemas-from-proto --single-branch git@github.com:norbjd/envoy.git
cd envoy/
docker run --rm -it -v $PWD:/app -w /app debian:sid-slim
```

*2. Prepare the environment* 

All commands from here are run inside the docker container.

```shell
apt update && apt install -y curl git gcc

# install bazelisk
curl -Lo /usr/local/bin/bazel https://github.com/bazelbuild/bazelisk/releases/download/v1.16.0/bazelisk-linux-amd64
chmod +x /usr/local/bin/bazel

# add a non-root user, required for other non-related bazel rules
adduser --disabled-password --gecos "" envoy
su envoy # connect with non-root user
```

*3. List new rules*

There are two ways to generate JSON schemas: using `rules_proto_grpc` or an aspect.

I don't know which one to use yet, so I've done both ways and let Envoy people decide.

```shell
# get the list of rules (using `rules_proto_grpc`)
$ bazel query 'kind("jsonschema_compile", //tools/protojsonschema:*)'
//tools/protojsonschema:envoy_config_bootstrap_v2
//tools/protojsonschema:envoy_config_bootstrap_v3

# get the list of rules (using an aspect)
bazel query 'kind("protojsonschema_rule", //tools/protojsonschema_with_aspects:*)'
//tools/protojsonschema_with_aspects:api_protojsonschema
```

*4. Build*

*4.1. Using `rules_proto_grpc`*

```shell
# example: generate the JSON schema for v3 bootstrap
$ bazel build //tools/protojsonschema:envoy_config_bootstrap_v3
...

# list generated files by the jsonschema_compile rule
$ ls -1 bazel-bin/tools/protojsonschema/envoy_config_bootstrap_v3
Admin.schema.json
Bootstrap.schema.json
ClusterManager.schema.json
CustomInlineHeader.schema.json
FatalAction.schema.json
LayeredRuntime.schema.json
Runtime.schema.json
RuntimeLayer.schema.json
Watchdog.schema.json
Watchdogs.schema.json

$ build all JSON schemas
bazel build $(bazel query 'kind("jsonschema_compile", //tools/protojsonschema:*)')
...
```

*4.2. Using an aspect*

```shell
# generate all JSON schemas
$ bazel build //tools/protojsonschema_with_aspects:api_protojsonschema
...

# list generated files by the api_protojsonschema rule (example here: v3 bootstrap)
$ ls -1 bazel-bin/external/envoy_api/envoy/config/bootstrap/v3/jsonschema
Admin.json
Bootstrap.json
ClusterManager.json
CustomInlineHeader.json
FatalAction.json
LayeredRuntime.json
Runtime.json
RuntimeLayer.json
Watchdog.json
Watchdogs.json
```

**Risk Level**

Low

**Testing**

I'm not really sure how to test this, other than just building the JSON schemas.

**Docs Changes**

/

**Release Notes**

Generate JSON schemas from proto files.

**Platform Specific Features**

/

**Fixes**

#13078, #13233, #13254.

### CODE REVIEW

**Review by phlax — COMMENTED** [review]
hi @norbjd thanks for working on this

i worked on it previously and almost got it working at that time, but other priorities took over

it was quite a while ago, so i dont rem exactly, but i think it was not implemented optimally

i think probably this requires use of a bazel aspect

File: bazel/jsonschema/BUILD
Comment (phlax) [inline bazel/jsonschema/BUILD:9]: we dont want to do this

its possible that this list could be derived by using a `genquery` but almost certainly this should be implemented using an aspect, which allows you to traverse dependencies, and visit individual targets

File: bazel/jsonschema/com_github_chrusty_protoc_gen_jsonschema_deps.bzl
Comment (phlax) [reply bazel/jsonschema/com_github_chrusty_protoc_gen_jsonschema_deps.bzl:2]: try:

```console
$ git grep go_repository
```

to view existing go repos

File: bazel/jsonschema_imports.bzl
Comment (phlax) [reply bazel/jsonschema_imports.bzl:16]: *all* dependencies should be registed in either `bazel/repository_locations.bzl` or `api/bazel/repository_locations.bzl`

you can use a commit sha and your own repo for testing, see existing examples

in this case it probably wants to live in `api/bazel/...`

**Comment by phlax** [comment]
i would suggest a first step to just focus on getting a working `protc-gen-jsonschema` binary in bazel

**Comment by dio** [comment]
Cool. This is awesome.

File: bazel/jsonschema_imports.bzl
Comment (phlax) [reply bazel/jsonschema_imports.bzl]: cool, i can look further after, but again trying to rem how i got the binary built previously

i agree that would be much better to upstream the bazel stuff, but we can also just create a bazel env in envoy if necessary

for now lets get an "ideal" bazel setup in your repo, and then figure it out from there

the bit i think i found difficult is using gazelle - ironically, as that is supposed to take the pain out of adding go deps

**Comment by phlax** [comment]
> I'm not sure to understand sorry confused

could you separate out adding protoc-gen-jsonschema to its own PR please - we can then ensure everything is working with that before looking at adding an aspect

File: bazel/jsonschema/BUILD
Comment (phlax) [reply bazel/jsonschema/BUILD]: as i said - i think this is _not_ the right way - we want to use an aspect

lets get the binary sorted first/separately

File: tools/protojsonschema_with_aspects/BUILD
Comment (phlax) [inline tools/protojsonschema_with_aspects/BUILD:1]: rather dont include _any_ use/rules/aspects etc in this rule - lets just get the binary working

File: bazel/jsonschema/BUILD
Comment (phlax) [reply bazel/jsonschema/BUILD]: np, re aspects, trying to rem how i did it previously, most likely not using them, but i think its the correct way, and im happy to help

**Comment by phlax** [comment]
so, @norbjd once #27661 lands we can figure out the next step

im also trying to think of the overall plan - im thinking the final outcome we want is that CI is automatically publishing the schemas somewhere

re use of an aspect - my niggling doubt is that im wondering if we ran `protoc` with the plugin against just the `Bootstrap` proto (and its dependencies) would that produce what we are after

i think when i tried this before i did it using an aspect, and then merged the files that it created (with various fixes/cleanups) - im wondering if that is necessary and whether this could actually be a lot simpler

**Comment by ravenblackx** [comment]
/wait

@phlax is on vacation at the moment so may be less responsive for a while.
Looks like there's some presubmits to be resolved in the meantime.

**Comment by KBaichoo** [comment]
maybe @keith might be able to take a look while phlax is out?

/assign @keith

**Comment by phlax** [comment]
> So now, there are two ways of generating the JSON schemas:

im a bit nervous about adding the grpc dep - managing existing grpc <> protobuf deps is already a challenge and in those cases we patch them and/or dont call their workspace init functions directly, so the setup is pretty complicated already

perhaps it wouldnt matter but i think if we have an option that doesnt require any additional deps - particularly related to these - then that would definitely be preferable

still catching up but will checkout the pr and do some local testing soon

**Comment by jmarantz** [comment]
Hm. @phlax is out? Is that right? @keith unassigned himself; not sure why.

Maybe @ravenblackx could take a look next? I'm not clear from looking at this PR if the next step is on the part of the maintainer or the CL author.

**Comment by phlax** [comment]
apologies @norbjd ive been a bit busy - ill look through this again as soon as i get chance

**Comment by ravenblackx** [comment]
Sorry, I didn't even notice this was assigned to me until right now. I don't think I'd be particularly helpful anyway though, I'm not familiar with IDE json-based validation, nor with bazel aspects (my earlier comment was just because I was on-call and sweeping through the things that were stuck - I guess at least I managed to help move it along a little!)

I'm going to unassign me to avoid causing confusion.

**Comment by htuch** [comment]
@phlax friendly ping on next steps for this PR.

**Comment by KBaichoo** [comment]
friendly ping @phlax

**Comment by phlax** [comment]
huge apology for delayed response - it just that this will take some time to review/test and ive been quite busy

ill try and look at this before the end of the week

/wait-any

**Comment by phlax** [comment]
> Kindly bumping this PR as it has been two months without activity 😇

again a huge apology - entirely my fault - the delay is that this takes some testing/iteration and i have been v busy

we have a batch of releases about to go out - once these are out of the way i _will_ priroritize this - apologies again @norbjd

**Comment by phlax** [comment]
/wait-any pending project release and further review

**Comment by chasepo** [comment]
@phlax, Do you have an timeframe on when you might have free cycles to tackle this review? I'm excited to be able to have some autocomplete support in VSCode for the yaml files, even if it's minimal at first! Thanks in advance.

**Comment by phlax** [comment]
apologies it again slipped my mind and the holiday season kicked in

no promises - but ill try and look at this next week

**Comment by ravenblackx** [comment]
@phlax has been pinged again through back channels more aggressively, so be assured this is still in the pipeline for attention, it's just a big thing and time is hard to find.

**Review by phlax — APPROVED** [review]
@norbjd apologies for taking so long to review this

lets land this and iterate, i think its a great step forward, lgtm

**Comment by norbjd** [comment]
Thanks @phlax! Just to remind, this PR showed 2 solutions because I was really unsure of the best one:

1. with proto_plugin (from rules_proto_grpc)
2. with aspects: see tools/api_proto_plugin/plugin.bzl and tools/protojsonschema_with_aspects

I'm not sure we needed both solutions to be merged to be fair. But, if you prefer one, tell me and I'll gladly make another PR to keep only one ;)

**Comment by phlax** [comment]
thanks @norbjd i vaguely remembered that to be the case - my preference is for the aspect approach - that was kinda what delayed me - wanting to test it

**Comment by norbjd** [comment]
Hello @davidfiala :wave: I remember that the generated JSON schema was not perfect for some fields (namely the filters). This is because we can't _easily_ express in the JSON schema that fields in `typed_config` depends on the name in the filter (e.g. if filter is `envoy.filters.network.http_connection_manager`, fields under `typed_config` will be `codec_type`, `stat_prefix`, ..., as defined [here](https://www.envoyproxy.io/docs/envoy/v1.30.2/api-v3/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto)). There might be a way to do it, but it will require some changes (see below).

Anyway first, just to explain why auto-completion on `filters` is not working: if you look into `Bootstrap.json`, the `filters` field is defined like:

```json
"filters": {
    "items": {
        "$ref": "#/definitions/envoy.config.cluster.v3.Filter"
    },
    "type": "array"
}
```

And the `#/definitions/envoy.config.cluster.v3.Filter` is defined (I've removed descriptions for readability) like:

```json
"envoy.config.cluster.v3.Filter": {
    "properties": {
        "name": {
            "minLength": 1,
            "type": "string"
        },
        "typed_config": {
            "properties": {
                "type_url": {
                    "type": "string"
                },
                "value": {
                    "type": "string",
                    "format": "binary",
                    "binaryEncoding": "base64"
                }
            },
            "additionalProperties": true,
            "type": "object"
        }
    },
    "additionalProperties": true,
    "type": "object",
    "title": "[#protodoc-title: Upstream filters]\n Upstream filters apply to the connections to the upstream cluster hosts."
}
```

As you can see, in `typed_config`, we just have `type_url` and `value` (an arbitrary string) + some "additional properties" (meaning _anything_: this is why we don't have auto-completion). To do things properly, we would have to:

1. generate the JSON schema for filters (I don't fully remember but I think these are not generated, correct me if I'm wrong)
2. include them in `Bootstrap.json`
3. configure properties in `envoy.config.cluster.v3.Filter` with something like [conditionals](https://json-schema.org/understanding-json-schema/reference/conditionals#ifthenelse), to represent things like "if filter name is `http_connection_manager`, then properties are `codec_type`, `stat_prefix`, etc". Maybe conditionals is not the appropriate solution though, but I can't think of something else right now

Note that this is just theoretical and might not be as easy as it seems! The [library](https://github.com/chrusty/protoc-gen-jsonschema) used here to convert protos to JSON schema probably lacks support for this.

> As an aside, it would be nice to generate these JSON artifacts as part of the release process so that others do not need to repeat the build steps.

Completely agree on this. As said in [this comment](https://github.com/envoyproxy/envoy/pull/27640#issuecomment-1568344767), "the final outcome we want is that CI is automatically publishing the schemas somewhere". But, alas, this has not been done yet. Merging this PR was just a first step to introduce JSON schemas, so there is still a lot of things to do! Unfortunately, I'm not sure if this is a priority today.

**Comment by phlax** [comment]
hi @norbjd @davidfiala just to clarify that the name is not so important wrt extensions - its more the `@type` field that determines what can be set

trying to rem whether i got this working when i first tried it - i thought i had - either way, i think it should be possible if non-trivial

would be great to get this working good enough to be usable, at that point we can think about publishing somewhere

**Pull request merged**
